Infra/PagingQuery.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Infra
+{
+    public class PagingQuery
+    {
+        int _pageSize;
+        int _page;
+
+        protected PagingQuery()
+        {
+            _pageSize = 25;
+            _page = 1;
+        }
+
+        public int PageSize
+        {
+            get { return _pageSize; }
+            set
+            {
+                if (value <= 0)
+                    throw new ArgumentOutOfRangeException("PageSize must be bigger than 0.");
+
+                if (value > 1000)
+                    throw new ArgumentOutOfRangeException("PageSize must be less than 1000.");
+
+                _pageSize = value;
+            }
+        }
+
+        public int Page
+        {
+            get { return _page; }
+            set
+            {
+                if (value <= 0)
+                    throw new ArgumentOutOfRangeException("Page must be bigger than 0.");
+
+                _page = value;
+            }
+        }
+    }
+
+    public interface IPage<T> : IEnumerable<T>
+    {
+        Pagination Pagination { get; }
+    }
+
+    class Page<T> : IPage<T>
+    {
+        public Page(Pagination pagination, IEnumerable<T> items)
+        {
+            Pagination = pagination;
+            Items = items;
+        }
+
+        public Pagination Pagination { get; }
+        IEnumerable<T> Items { get; }
+        public IEnumerator<T> GetEnumerator() => Items.GetEnumerator();
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+    }
+
+    public static class Page
+    {
```

Comment:
@johnnyoshika What do you think about inserting the following here:

        public static IPage<TResult> ToPage<T, TResult>(this IQueryable<T> source, PagingQuery query, Func<T, TResult> selector) =>
            source
                .AsPage(query, selector)
                .Load();

        public static IPage<T> ToPage<T>(this IQueryable<T> source, PagingQuery query) =>
            source
                .AsPage(query)
                .Load();

        static IPage<T> Load<T>(this IPage<T> source) =>
            new Page<T>(
                source.Pagination,
                source.ToArray());

Noting being braked, but we will have more flexibility with supporting AsPage and ToPage the same time:
* `As` version is a wrapper (adapter) over remote data source, so data being queried on enumeration (more than one time possibly).
*  `To` version loads and caches all in memory at once by creating a detached object with all the state of original one.

We have it everywhere in .NET: 
* `IList<T>.AsReadonly()`
* `IEnumerable<T>.ToList()`